src/System.Resources.Analyzers/Core/MarkAssembliesWithNeutralResourcesLanguage.cs

```diff
@@ -13,26 +16,124 @@ public abstract class MarkAssembliesWithNeutralResourcesLanguageAnalyzer : Diagn
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/bb385967.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
+        protected abstract void RegisterAttributeAnalyzer(CompilationStartAnalysisContext context, Action onResourceFound);
+
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+            analysisContext.EnableConcurrentExecution();
+
+            var hasResource = false;
+            analysisContext.RegisterCompilationStartAction(cc =>
+            {
+                RegisterAttributeAnalyzer(cc, () => hasResource = true);
+
+                cc.RegisterCompilationEndAction(ce =>
+                {
+                    // there is nothing to do.
+                    if (!hasResource)
+                    {
+                        return;
+                    }
+
+                    AttributeData data;
+                    if (TryCheckNeutralResourcesLanguageAttribute(ce, out data))
+                    {
+                        // attribute already exist
+                        return;
+                    }
+
+                    if (data != null)
+                    {
+                        // we have the attribute but its doing it wrong.
+                        ce.ReportDiagnostic(data.ApplicationSyntaxReference.GetSyntax(ce.CancellationToken).CreateDiagnostic(Rule));
+                        return;
+                    }
+
+                    // attribute just don't exist
+                    ce.ReportDiagnostic(Diagnostic.Create(Rule, Location.None));
+                });
+            });
+        }
+
+        protected static bool CheckDesignerFile(SyntaxTree tree)
+        {
+            return tree.FilePath?.IndexOf(Designer, StringComparison.OrdinalIgnoreCase) > 0;
+        }
+
+        protected static bool CheckResxGeneratedFile(SemanticModel model, SyntaxNode attribute, SyntaxNode argument, CancellationToken cancellationToken)
+        {
+            if (!CheckDesignerFile(model.SyntaxTree))
+            {
+                return false;
+            }
+
+            var generatedCode = WellKnownTypes.GeneratedCodeAttribute(model.Compilation);
+            if (model.GetSymbolInfo(attribute, cancellationToken).Symbol?.ContainingType?.Equals(generatedCode) != true)
```

Comment:
Ah that was a syntactic check for better perf? It's confusing since they are far apart. If you write this as a symbol analyzer you can move them to the same method.